### ISSUE TITLE
Watch not working with docker

### ISSUE THREAD

**Issue by azhar1038** [esbuild-kit/tsx#266]
### Bug description

I have created one file named `main.ts` and am using `tsx watch ./main.ts` as my start script to watch the file. Locally everything is working fine (Node 20) but when I try to run in a docker container, it does not watch changes in `main.ts`.
 I have mounted my local files to container.
![Screenshot 2023-07-12 220303](https://github.com/esbuild-kit/tsx/assets/42082172/9f264606-929e-483c-8963-3d1486f5cceb)

I was expecting it to reload when I make changes in `main.ts` locally.

Nothing happens. `tsx` does not reload anything.

I did not get any error

### Reproduction

This is my `main.ts`
```ts
console.log('Hello world!');
```

This is my `Dockerfile` (I have tried with node:alpine too)
```
FROM node:18-alpine
WORKDIR /app
COPY package.json .
COPY package-lock.json .
RUN npm install
COPY . .
CMD [ "npm", "run", "start"]
```

This is the `docker-compose.yml`
```
version: '3.8'
services:
  tsx-test:
    build: .
    volumes:
      - .:/app
      - nodeModules:/app/node_modules

volumes:
  nodeModules:
```

### Environment

```shell
System:
    OS: Windows 10 10.0.22621
    CPU: (8) x64 Intel(R) Core(TM) i5-9300H CPU @ 2.40GHz
    Memory: 1.23 GB / 7.85 GB
  Binaries:
    Node: 20.1.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 3.5.1 - C:\Program Files\nodejs\yarn.CMD
    npm: 9.6.4 - C:\Program Files\nodejs\npm.CMD
    pnpm: 8.6.7 - C:\Program Files\nodejs\pnpm.CMD
  npmPackages:
    tsx: ^3.12.7 => 3.12.7 

BTW, I am using Windows 11 not 10
```

### Can you work on a fix?

- [ ] I’m interested in opening a pull request to address this issue.

**Comment by privatenumber** [comment]
I don't use Docker but if you want to investigate and open a PR, feel free.

BTW, the watcher used is chokidar, so perhaps it's a bug in that.

**Comment by azhar1038** [comment]
Okay, I will check and see if I can find the issue.

**Comment by azhar1038** [comment]
Hi @privatenumber,
I found the problem!

It is mentioned in [Chokidar documentation](https://github.com/paulmillr/chokidar#performance) that we need to set `usePolling` to `true` to be able to watch files over a network. Problem is setting it to true can cause performance issue.

Then I checked Nodemon to see how they have handled this and found that they use [a flag `--legacy-watch` or `-L`](https://github.com/remy/nodemon#application-isnt-restarting) to set `usePolling` to `true`.

Shall I add a similar flag to set `usePolling` in tsx?

**Comment by privatenumber** [comment]
Thanks for the investigation.

Ideally no new flags... Node.js already added `watch` mode, and I don't want to end up maintaining a watch feature as it's not the main point of this project.

How does Node.js's watch handle this?

**Comment by azhar1038** [comment]
Thanks for the reply.
I have checked node experimental watch mode already and it does not handle this either.

**Comment by privatenumber** [comment]
Okay, lets add a --poll flag that accepts a interval value.

**Comment by privatenumber** [comment]
Actually, it may not be necessary if we change watchers: https://github.com/esbuild-kit/tsx/issues/246

Does Parcel's watcher compatible with network files?

**Comment by azhar1038** [comment]
I am not sure. I will check.

**Comment by azhar1038** [comment]
I checked @parcel/watcher and it also does not work with Docker on windows.

In fact, no watcher can actually handle this. The problem is in Windows, Docker uses WSL2 which uses it own file system (ext4 for Ubuntu) instead of windows NTFS.
So, when I change something locally in windows, change happens in NTFS but container running in ext4 never gets notified of the change because of this bug in WSL: [[WSL2] File changes made by Windows apps on Windows filesystem don't trigger notifications for Linux apps](https://github.com/microsoft/WSL/issues/4739).

So, until that issue is fixed (which will definitely take a long time) we have a few options:
- Move the project to linux file system and do everything from there
- Use polling (parcel does not support this)
- Switch to linux :)

I think switching to @parcel/watcher is actually a better long term option as Chokidar is a bit less active now.

**Comment by matt-sanders** [comment]
I'm having a similar issue here which is potentially related. If not, let me know and I can open a new issue. But when using this with Docker on an M2 Mac, changes to files are not being picked up. Interestingly if I use nodemon it works as expected. [Example repo here](https://github.com/matt-sanders/docker-macos-reload-error).

**Comment by azhar1038** [comment]
Both tsx and nodemon uses [Chokidar](https://www.npmjs.com/package/chokidar) to watch file changes. So, I don't think it is same issue.
However, it is possible that nodemon is setting `usePolling` or `useFsEvents` for mac somehow.
Unfortunately I don't have access to any mac, can you please try to see if chokidar is directly working in docker or not? If it does then there is some other issue in tsx.

**Comment by matt-sanders** [comment]
Can confirm that Chokidar works fine without TSX. I've pushed another script to the above repo as an example.

**Comment by azhar1038** [comment]
Then your issue is probably related to #246

**Comment by matt-sanders** [comment]
Hmm, we're actually using Node 18. Interestingly it seems to work with colima but not with Docker Desktop. For now I'm just either continuing with colima or if I really need it, I use Chokidar's polling.

**Comment by adam-beck** [comment]
I have a similar issue but I don't know if it's related. I'm using [Compose Watch](https://docs.docker.com/compose/file-watch/) instead of volumes. For the first few file changes, it works great. But after like 4-5 it just stops. I'm not sure how to debug either.

In the meantime -- and the reason I'm posting this comment -- is that I resolved to using polling with chokidar. I'm still using `tsx` but I'm also passing in the environment variable `CHOKIDAR_USEPOLLING=true` to my container that is running `tsx`. Seems to "fix" the issue. Well, as good as any other alternative at the moment.

**Comment by vtonchev** [comment]
I am trying to use the same configuration as you @adam-beck , however no results. Is there something else aside from CHOKIDAR_USEPOLLING=true to be done. Thanks

**Comment by azhar1038** [comment]
@adam-beck Yes, I just tried compose watch and it is behaving weirdly. When I use:
- experimental native `--watch`: changes are reflected only once
- `tsx watch`: changes are reflected around 4 times and then stops
- `nodemon`: works fine even without polling!
- `tsx` with `CHOKIDAR_USEPOLLING=true` works fine as expected

Not sure why tsx watch stops working after some time.
Here is my `compose.yml` in case anyone is interested.
```yml
# compose.yml

services:
  server:
    build:
      context: .
    ports:
      - 3000:3000
    environment:
      - CHOKIDAR_USEPOLLING=true
    develop:
      watch:
        - action: sync
          path: ./src
          target: /app/src
        - action: rebuild
          path: package.json
          target: /app
```

**Comment by adam-beck** [comment]
I wrote a simple script to determine if chokidar was picking up the changes via compose watch

```js
const chokidar = require('chokidar');

// One-liner for current directory
chokidar.watch('.').on('all', (event, path) => {
  console.log(event, path);
});
```

Every change seemed to be picked up. So I'm starting to think it's something within tsx itself. Granted, this is a JS script but I don't think that would matter.

**Comment by privatenumber** [comment]
I'm likely going to migrate to using Parcel watcher in the near future.

Can one of you give me a reproduction in a GitHub Action environment (or something similar) where I can reproduce this bug? This would help me cover this use-case.

**Comment by insanity54** [comment]
Hi hi, thank you for tsx. I've been using it for a month or two and I really enjoy it.

I've been bit by this issue so I wanted to see if there's anything I can do to help figure it out. I've made a reproduction repo along with a couple Github actions to test the `--watch` behavior. I'm not sure it's correct but maybe it's a start.

https://github.com/insanity54/tsx-issue-266/blob/master/.github/workflows/tests.yaml

**Comment by alexmorleyfinch** [comment]
For anyone reading who just wants this to work, set environment variable `CHOKIDAR_USEPOLLING=1`

It's already mentioned above but I wanted to confirm it's a quick work around that works for most cases

**Comment by psibean** [comment]
I previously had this working on Windows in December last year.

I had this Dockerfile:

### jsmud-api

```
FROM node:20-alpine
WORKDIR /app
COPY . .
RUN npm install
EXPOSE 4000
CMD ["npm",  "run", "dev", "--host 0.0.0.0"]
```
Where dev script:

```
"dev": "tsx --watch --inspect=0.0.0.0 ./src/index.js",
```

And I had these `tsconfig` `watchOptions`:

```
  "watchOptions": {
    "watchFile": "dynamicPriorityPolling",
    "watchDirectory": "dynamicPriorityPolling",
  }
```
I had these volume mounts in the docker-compose:
```
        volumes:
            - /app/node_modules
            - ./jsmud-api:/app
```

At the time I was using `tsx@4.19.1` and TypeScript `5.6.2`.

**This used to be working for me.** It no longer is, and I'm not sure why.

The main thing that's changed in my environment is that the main user account is no longer an Administrator, where it used to be back then. I'm wondering if there's some funky Windows permissions preventing this from working like it used to.

Because it's not just `tsx` watch that has stopped working for me in this way. In the same project I have a React SPA generated by Vite, and I have the Vite config set with watch to usePolling. It used to work as well, but no longer does.
